mahidhar/segtree.cpp

```diff
@@ -0,0 +1,106 @@
+//This is a generic implementation of the segment tree without lazy propagation
+//This solves range sum query with updates
+//For other problems the implementation of the util class should be changed accordingly.
+
+#include <iostream>
+
+struct sTreeNode{
+	int x;
+};
+
+typedef struct sTreeNode snode;
+
+class util{
+public:
+	//This is the return value if the range is out of bounds
+	snode outValue;
+
+	util(){
+		outValue.x = 0;
+	}
+	snode getLeaf(int a){
+		snode ans;
+		ans.x = a;
+		return ans;
+	}
+	snode combine(const snode &a,const snode &b){
+		snode ans;
+		ans.x = a.x+b.x;
+		return ans;
+	}
+	snode getVal(const snode &a){
+		snode ans;
+		ans.x = a.x;
+		return ans;
+	}
+};
+
+//IMPORTANT:: Assumes array is 1 indexed.
+class segtree{
+public:
+	int SIZE;
```

Comment:
const int SIZE;

Reply:
It ideally must be unsigned int but okay for now. Also member variables must be private.